### PR TITLE
Various improvements

### PR DESCRIPTION
**Pull request by blackheaven**
Hello,

I have seen your project during ZuriHac, and I have been quite busy in the mean time, I have finally find some for another project of mine and I make use of it.

This PR packs up few improvements:

 * The naming was incorrect (the cabal file was not the right one and module names was not matching)
 * It introduced `stack`
 * It adds Hspec support

I also take the opportunity to do some suggestions:

 * use `hpack`
 * publish it on hackage

Do not hesitate to ask for changes.

### CODE REVIEW

**Comment by alexbiehl** [comment]
Thank you Gautier! This is looking great!

File: test/TestContainers/HspecSpec.hs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+containers1 = do
+  _ <- run redis defaultContainerRequest
+  pure ()
+
+
+main :: IO ()
+main = hspec spec_all
+
+spec_all :: Spec
+spec_all = around (withContainers containers1) $
```
Comment (alexbiehl) [inline test/TestContainers/HspecSpec.hs:20]: Hspec integration looks much more convenient than Tasty's! Cool!

**Comment by alexbiehl** [comment]
Regarding release on hackage, do you have a good sense of what still needs to happen for a 0.1 release? Here are a few random thoughts 

* I suggest at least getting rid of all the undefines (either by implementing or getting rid of them) should be resolved. 
* Also I think we should make run wait for the container to come up and out the WaitUntilReady check into the ContainerRequest
* Also put ToImage into the ContainerRequest to have it fully self-contained

Let me know if you have suggestions and/or want to spend time on anything so that we can divide the workload. 

Also thanks again for the PR much appreciated!

File: src/TestContainers/Hspec.hs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+--
+-- containers :: MonadDocker m => m Boolean
+-- containers = do
+--   _redis <- TestContainers.run TestContainers.redis TestContainers.defaultContainerRequest
+--   _kafka <- TestContainers.run TestContainers.kafka TestContainers.defaultContainerRequest
+--   pure True
+--
+-- example :: Spec
+-- example =
+--   around (withContainers containers) $ describe "Example tests"
```
Comment (alexbiehl) [inline src/TestContainers/Hspec.hs:36]: I think the right hand side must be a lambda, no?

Comment (alexbiehl) [reply src/TestContainers/Hspec.hs:36]: Ah, didn't know how around works!

**Comment by blackheaven** [comment]
Cool, yes, nice ideas, I will have few more:

 * Add a wrapper around `docker inspect`
 * Add basic information fetching (for example, [here is how I fetch the IP](https://github.com/blackheaven/amqp-streamly/blob/f37e729a4e5b0127bd7ae41669b2321b66ef889c/test/Network/AMQP/StreamlySpec.hs#L77))

**Comment by alexbiehl** [comment]
Makes sense! Also I think it makes sense to add a WaitUntilReady that blocks until the container listens on some defined port.

**Comment by alexbiehl** [comment]
Hey @blackheaven, I addressed all the points. Do you mind taking a look at the master branch and see if there is anything obviously missing?

**Comment by blackheaven** [comment]
Great, two last points:

 - rename `id` to `containerId`
 - have a log-based wait strategy, see (this code)[https://github.com/blackheaven/hermes/blob/master/test/Control/Hermes/RabbitMQUtils.hs#L111]

IMO you are good to publish it on hackage.

**Comment by alexbiehl** [comment]
> * rename id to containerId

Do you think I should prefix every field of `Container` with `container` or just `id`? 

> * have a log-based wait strategy

FWIW we have `waitForLogLine`: 

```
wairForLogLine Stdout (" completed with " `LazyText.isInfixOf`)
```

**Comment by alexbiehl** [comment]
Your example would boil down to something like 

```
containers :: MonadDocker m => m (Channel, Channel)
containers = do
  let (login, password) = ("guest", "guest")

  container <- run $ containerRequest (fromTag "rabbitmq:3.8.4")
    & setWaitingFor (waitForLogLine (" completed with " `LazyText.isInfixOf`))

  connection <- liftIO $ openConnection (ip container) "/" login password

  (,)
    <$> liftIO (openChannel connection)
    <*> liftIO (openChannel connection)
```

**Comment by blackheaven** [comment]
> Do you think I should prefix every field of `Container` with `container` or just `id`?

every field

> FWIW we have `waitForLogLine`:

I have missed it, that's awesome!

**Comment by blackheaven** [comment]
A last set of remarks:

 * dependencies bounds does not match to any stable stack stable release
 * relying on `lens` seems a bit heavy for your usecase

**Comment by alexbiehl** [comment]
> dependencies bounds does not match to any stable stack stable release

What does that mean? How would I check against those?


> relying on lens seems a bit heavy for your usecase

I agree. It made for such beatiful traversal of the JSON stuff.

**Comment by blackheaven** [comment]
> > dependencies bounds does not match to any stable stack stable release
> 
> What does that mean? How would I check against those?

Just configure stack with `lts-16.8` you'll get:

```
In the dependencies for testcontainers-0.1.0.0:
    lens-4.18.1 from stack configuration does not match >=4.19.2 && <5  (latest matching version
                is 4.19.2)
    network-3.1.1.1 from stack configuration does not match >=2.8.0 && <2.9  (latest matching
                    version is 2.8.0.1)
    unliftio-core-0.1.2.0 from stack configuration does not match >=0.2.0 && <0.3  (latest
                          matching version is 0.2.0.1)
needed due to amqp-streamly-0.2.0 -> testcontainers-0.1.0.0

Some different approaches to resolving this:

  * Set 'allow-newer: true'
    in /home/black/.stack/config.yaml to ignore all version constraints and build anyway.

  * Recommended action: try adding the following to your extra-deps
    in /home/black/programmation/haskell/amqp-streamly/stack.yaml:

- lens-4.19.2@sha256:f0570ca0e4624d9daa57256f00f514e87be155401583fa1665faba47376c63a7,16053
- network-2.8.0.1@sha256:0f165dffa752d8cde30c2bde86f80609c4f1dc5eeb3182d593041f97839c5b3b,3011
- unliftio-core-0.2.0.1@sha256:9b3e44ea9aacacbfc35b3b54015af450091916ac3618a41868ebf6546977659a,1082

Plan construction failed.
```

> > relying on lens seems a bit heavy for your usecase
> 
> I agree. It made for such beatiful traversal of the JSON stuff.

I thought there were a `optics-aeson` library :/

**Comment by alexbiehl** [comment]
Are you able to provide better bounds? I don't have stack installed and all. If you know how to get to it it would be cool if you provided them. I will look at it eow else. 

Oh, optics-aeson sounds promising. Will switch to that, thanks!

**Comment by blackheaven** [comment]
Yes and yes, I am on it (better bounds and `optics-aeson`)

**Pull request merged**